tensorflow_addons/losses/center_loss.py

```diff
@@ -0,0 +1,86 @@
+# Copyright 2019 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Implements center loss."""
+
+
+import tensorflow as tf
+from tensorflow_addons.utils.types import FloatTensorLike
+from typeguard import typechecked
+
+
+@tf.keras.utils.register_keras_serializable(package="Addons")
+class CenterLoss(tf.keras.layers.Layer):
+    """Computes the center loss.
+
+    See: https://ydwen.github.io/papers/WenECCV16.pdf
+    The loss was designed to develop an effectively improve the discriminative
+    power of the deep learned features. In fact, the key is to minimize the
+    intra-class variations while keeping the features of different classes
+    separable. Usually combining with softmax losss to jointly supervise the
+    CNNs to have a better result.
+    We expect labels `y_pred` must be the output of the last fully connected
+    layer, which is 2-D float `Tensor` of l2 normalized embedding vectors and
+    `y_true` to be provided as 1-D integer `Tensor` with shape [batch_size]
+    of multi-class integer labels.
+
+    Args:
+      name: Optional name for the op.
+      reduction: (Optional) Type of `tf.keras.losses.Reduction` to apply to
+        loss. Default value is `SUM_OVER_BATCH_SIZE`.
+
+    Returns:
+      Tensor containing calculated center loss.
+    """
+
+    @typechecked
+    def __init__(
+        self, alpha: FloatTensorLike = 0.2, name: str = "center_loss", **kwargs
+    ):
+        super().__init__(name=name, **kwargs)
+        self.alpha = alpha
+
+    def call(self, y_true, y_pred):
+        return self.loss_fn(y_true, y_pred, self.alpha)
+
+    def loss_fn(self, labels, feature, alpha):
+        labels = tf.convert_to_tensor(labels)
+        feature = tf.convert_to_tensor(feature)
+        len_features = feature.shape[-1]
+        num_classes = labels.shape[-1]
+        centers = tf.Variable(
+            tf.constant(0.0, shape=[num_classes, len_features]),
+            name="centers",
+            dtype=tf.float32,
+        )
```

Comment:
Maybe it is better to move variable creation into `build`. Also please use `add_weight` to create variables when subclassing `Layer`

https://www.tensorflow.org/api_docs/python/tf/keras/layers/Layer#add_weight

Reply:
@WindQAQ, Here we are using two shapes `labels.shape[-1]` and `feature.shape[-1]` to define `centers` variable. But in `build` method, we can only get one of them, explicitly saying only `labels.shape`. So, to change that, I have to pass `num_classes` explicitly in `__init__`. So, should we do that or should we keep this?